app/main.py

```diff
@@ -4,31 +1,64 @@
+        self.row = row
+        self.column = column
+        self.is_alive = is_alive
 
 class Ship:
-    def __init__(self, start, end, is_drowned=False):
-        # Create decks and save them to a list `self.decks`
-        pass
+    def __init__(self, start: Tuple[int, int], end: Tuple[int, int],
+                 is_drowned: bool = False) -> None:
+        self.decks: List[Deck] = []
+        self.is_drowned = is_drowned
+        self.create_decks(start, end)
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+    def create_decks(self, start: Tuple[int, int],
+                     end: Tuple[int, int]) -> None:
+        start_row, start_col = start
+        end_row, end_col = end
+        if start_row == end_row:
+            for col in range(start_col, end_col + 1):
+                self.decks.append(Deck(start_row, col))
+        elif start_col == end_col:
+            for row in range(start_row, end_row + 1):
+                self.decks.append(Deck(row, start_col))
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
 
+    def fire(self, row: int, column: int) -> bool:
+        deck = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.is_alive = False
+        if all(not d.is_alive for d in self.decks):
+            self.is_drowned = True
+            return True
+        return False
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: List[Tuple[Tuple[int, int], Tuple[int, int]]])\
+            -> None:
+        self.field: Dict[Tuple[int, int], Ship] = {}
+        self.ships: List[Ship] = []
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def fire(self, location: Tuple[int, int]) -> str:
+        row, col = location
+        if (row, col) in self.field:
+            ship = self.field[(row, col)]
+            if ship.fire(row, col):
+                if ship.is_drowned:
+                    return "Sunk!"
+                return "Hit!"
+            return "Hit!"
```

Comment:
This block of code is a bit confusing. If `ship.fire(row, col)` returns `True`, then `ship.is_drowned` will always be `True` because it is set in the `fire` method of the `Ship` class. So the check `if ship.is_drowned:` is redundant and can be removed. This is a code quality issue.